File: algorithms/minimum_arrow_shots.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# 452. Minimum Number of Arrows to Burst Balloons
# https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/

import unittest
from typing import List

class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        points.sort(key=lambda x: x[1])
        shots = 1
        right = points[0][1]
        for p in points:
            if p[0] <= right:
                continue
            else:
                right = p[1]
                shots += 1
        return shots

class TestSolution(unittest.TestCase):
    def test_findMinArrowShots(self):
        sol = Solution()
        self.assertEqual(sol.findMinArrowShots([[10, 16], [2, 8], [1, 6], [7, 12]]), 2)
        self.assertEqual(sol.findMinArrowShots([[1, 2], [3, 4], [5, 6], [7, 8]]), 4)
        self.assertEqual(sol.findMinArrowShots([[1, 2], [2, 3], [3, 4], [4, 5]]), 2)

if __name__ == '__main__':
    unittest.main()

File: algorithms/minimum_arrow_shots.cpp
/*
 * =====================================================================================
 *
 *       Filename:  minimum_arrow_shots.cpp
 *
 *    Description:  452. Minimum Number of Arrows to Burst Balloons
 *
 *        Version:  1.0
 *        Created:  01/31/2024 09:56:04
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  xianfeng.zhu@gmail.com
 *   Organization:
 *
 * =====================================================================================
 */

#include <algorithm>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"

using std::vector;

class Solution {
 public:
  int findMinArrowShots(vector<vector<int>>& points) {
    if (points.size() == 0) {
      return 0;
    }

    std::sort(points.begin(), points.end(),
              [](const vector<int>& a, const vector<int>& b) { return a[0] < b[0]; });

    int arrows = 1;
    int a = INT_MIN;
    int b = INT_MAX;
    for (const auto& p : points) {
      a = std::max(a, p[0]);
      b = std::min(b, p[1]);
      if (a > b) {
        arrows++;
        a = p[0];
        b = p[1];
      }
    }

    return arrows;
  }
};

TEST(Solution, findMinArrowShots) {
  vector<std::pair<vector<vector<int>>, int>> cases = {
      std::make_pair(vector<vector<int>>{{10, 16}, {2, 8}, {1, 6}, {7, 12}}, 2),
      std::make_pair(vector<vector<int>>{{1, 2}, {3, 4}, {5, 6}, {7, 8}}, 4),
      std::make_pair(vector<vector<int>>{{1, 2}, {2, 3}, {3, 4}, {4, 5}}, 2),
  };
  for (auto& c : cases) {
    EXPECT_EQ(Solution().findMinArrowShots(c.first), c.second);
  }
}
